File: Server/src/main/java/jk_5/nailed/server/command/CommandReload.java
package jk_5.nailed.server.command;

import io.netty.channel.ChannelFuture;
import jk_5.nailed.api.NailedAPI;
import jk_5.nailed.api.concurrent.Callback;
import jk_5.nailed.api.concurrent.scheduler.NailedRunnable;
import jk_5.nailed.api.map.Map;
import jk_5.nailed.api.map.MappackLoader;
import jk_5.nailed.ipc.IpcManager;
import jk_5.nailed.permissions.NailedPermissionFactory;
import net.minecraft.command.ICommandSender;
import net.minecraft.util.ChatComponentText;
import net.minecraft.util.EnumChatFormatting;
import net.minecraft.util.IChatComponent;
import net.minecraftforge.permissions.api.PermBuilderFactory;
import net.minecraftforge.permissions.api.PermissionsManager;

import java.util.List;

/**
 * No description given
 *
 * @author jk-5
 */
public class CommandReload extends NailedCommand {

    public CommandReload() {
        super("reload");
    }

    @Override
    public void processCommandWithMap(final ICommandSender sender, Map map, String[] args) {
        if(args.length != 1){
            this.printUsage(sender);
        }else{
            String rule = args[0];
            if(rule.equals("permissions")){
                PermBuilderFactory<?> factory = PermissionsManager.getPermFactory();
                if(factory instanceof NailedPermissionFactory){
                    NailedPermissionFactory permFactory = (NailedPermissionFactory) factory;
                    permFactory.readConfig();
                    ChatComponentText component = new ChatComponentText("Reloaded permissions!");
                    component.getChatStyle().setColor(EnumChatFormatting.GREEN);
                    sender.addChatMessage(component);
                }
            }else if(rule.equals("mappacks")) {
                NailedAPI.getMappackLoader().loadMappacks(new Callback<MappackLoader>() {
                    @Override
                    public void callback(MappackLoader obj) {
                        IChatComponent component = new ChatComponentText("Successfully loaded " + obj.getMappacks().size() + " mappacks");
                        component.getChatStyle().setColor(EnumChatFormatting.GREEN);
                        sender.addChatMessage(component);
                    }
                });
            }else if(rule.equals("ipc")){
                NailedAPI.getScheduler().runTaskAsynchronously(new NailedRunnable() {
                    @Override
                    public void run(){
                        ChannelFuture future = IpcManager.instance().close();
                        if(future != null){
                            future.syncUninterruptibly();
                        }
                        IpcManager.instance().start();
                    }
                });
            }else{
                this.printUsage(sender);
            }
        }
    }

    private void printUsage(ICommandSender sender){
        IChatComponent comp = new ChatComponentText("Usage: /reload <permissions/mappacks>");
        comp.getChatStyle().setColor(EnumChatFormatting.RED);
        sender.addChatMessage(comp);
    }

    @Override
    public List<String> addAutocomplete(ICommandSender sender, String[] args) {
        if(args.length == 1){
            return getOptions(args, "permissions", "mappacks", "ipc");
        }else{
            return null;
        }
    }
}

File: Server/src/main/scala/jk_5/nailed/server/command/CommandReload.scala
package jk_5.nailed.server.command

import net.minecraft.command.ICommandSender
import jk_5.nailed.api.map.Map
import jk_5.nailed.util.ScalaUtils._
import jk_5.nailed.util.ScalaCallback._
import net.minecraftforge.permissions.api.PermissionsManager
import jk_5.nailed.permissions.NailedPermissionFactory
import net.minecraft.util.{ChatComponentText, EnumChatFormatting}
import jk_5.nailed.api.NailedAPI
import jk_5.nailed.ipc.IpcManager

/**
 * No description given
 *
 * @author jk-5
 */
object CommandReload extends ScalaCommand {

  val name = "reload"
  val usage = "/reload <mappacks|permissions|ipc> - Reloads the specified module"

  override def processCommandWithMap(sender: ICommandSender, map: Map, args: Array[String]){
    if(args.length != 1) throw new WrongUsageException("/reload <mappacks|permissions|ipc>")
    else{
      caseInsensitiveMatch(args(0)){
        case "permissions" =>
          PermissionsManager.getPermFactory match {
            case f: NailedPermissionFactory =>
              f.readConfig()
              val component = new ChatComponentText("Reloaded permissions!")
              component.getChatStyle.setColor(EnumChatFormatting.GREEN)
              sender.addChatMessage(component)
            case _ =>
          }
        case "mappacks" =>
          NailedAPI.getMappackLoader.loadMappacks(wrapCallback(l => {
            val comp = new ChatComponentText("Successfully loaded " + l.getMappacks.size() + " mappacks")
            comp.getChatStyle.setColor(EnumChatFormatting.GREEN)
            sender.addChatMessage(comp)
          }))
        case "ipc" =>
          NailedAPI.getScheduler.runTaskAsynchronously {
            case _ =>
              val future = IpcManager.instance().close()
              if(future != null) future.syncUninterruptibly()
              IpcManager.instance().start()
          }
        case _ => throw new WrongUsageException("/reload <mappacks|permissions|ipc>")
      }
    }
  }

  override def addAutocomplete(sender: ICommandSender, args: Array[String]) = if(args.length == 1) getOptions(args, "permissions", "mappacks", "ipc") else null
}
